### ISSUE TITLE
Flow is significantly broken

### ISSUE THREAD

**Issue by PeterKottas** [react-page/react-page#561]
Reporting this issue in hope of finding reasonable solution to this.
At the moment, there's an archaic version on flow-bin used in this project. When I updated it, it came up with 1000+ errors.
The flow as is currently used forces developers to adhere to "fake rules"
For example instead of providing props and state types via generic parameters into the Component<PropsType, StateType>, you have to initialize props: PropsType and state: StateType which is just not how it should be.
There's also lot of fake positives with possibly undefined, the type safety doesn't work well in jsx.

Another thing is the pipeline, right now, on my machines:
```
$ flow check --show-all-errors
Found 0 errors
Done in 269.32s
```

That's 269 seconds to typecheck. I come from typescript and I never even noticed it actually has to do type-checking, it's instantaneous. But this is a big problem. I've spend last 3 hours just fixing some stuff and waiting ages to see if it helped. 

What can we do about this?

**Comment by arekkas** [comment]
Yeah there are several issues with flow. The first is that the typedefs are not used or vastly outdated. Same goes for flow-bin. The community has moved on to TypeScript which sees many more community contributions. One (work-intense) solution is to rewrite the lib in TypeScript. Another is to fix the aforementioned issues.

**Comment by PeterKottas** [comment]
I am actually tempted to port it to typescript. I have gained significant knowledge of this lib and while I can't say it would be easy, it's definitely doable. Cause the thing is you're totally right about the community. Working with typescript (our whole code base is in that) has been a pleasure. And also it's quite fast. That's why even if we managed to fix these problems, we would still end up with type-system that imho is worse than typescript.
